tests/compas_timber/test_nesting.py

```diff
@@ -0,0 +1,477 @@
+    assert not stock.section_matches(beam4)
+
+
+def test_beam_addition_and_capacity_tracking():
+    """Test adding beams and tracking used length/waste."""
+    stock = Stock(6000, (60, 120))
+    stock.cutting_tolerance = 1.0
+
+    # Add first beam
+    beam1 = Beam(frame=Frame.worldXY(), length=2000, width=60, height=120)
+    stock.add_beam(beam1)
+
+    expected_utilized_length1 = beam1.blank_length + stock.cutting_tolerance
+    expected_waste1 = stock.length - beam1.blank_length - stock.cutting_tolerance
+    expected_frame1 = Frame.worldXY()
+
+    assert len(stock.beam_data) == 1
+    assert str(beam1.guid) in stock.beam_data
+    assert stock.beam_data[str(beam1.guid)]["length"] == 2000
+    assert stock.beam_data[str(beam1.guid)]["frame"] == expected_frame1
+    assert stock.utilized_length == expected_utilized_length1
+    assert stock.waste == expected_waste1
+
+    # Add second beam
+    beam2 = Beam(frame=Frame.worldXY(), length=1500, width=60, height=120)
+    stock.add_beam(beam2)
+
+    expected_utilized_length2 = expected_utilized_length1 + beam2.blank_length + stock.cutting_tolerance
+    expected_waste2 = expected_waste1 - beam2.blank_length - stock.cutting_tolerance
+    expected_frame2 = Frame.worldXY()
+    expected_frame2.point.x = beam1.blank_length + stock.cutting_tolerance
+
+    assert len(stock.beam_data) == 2
+    assert str(beam2.guid) in stock.beam_data
+    assert stock.beam_data[str(beam2.guid)]["length"] == 1500
+    assert stock.beam_data[str(beam2.guid)]["frame"] == expected_frame2
+    assert stock.utilized_length == expected_utilized_length2
+    assert stock.waste == expected_waste2
+
+
+def test_beam_fitting_validation():
+    """Test beam fitting validation and error handling."""
+    stock = Stock(6000, (60, 120))
+    cutting_tolerance = 0.0
+    stock.cutting_tolerance = cutting_tolerance
+
+    # Test beam that fits
+    beam1 = Beam(frame=Frame.worldXY(), length=2000, width=60, height=120)
+    assert stock.can_fit_beam(beam1)
+
+    # Add the beam
+    stock.add_beam(beam1)
+    assert stock.waste == 4000
+
+    # Test beam that exactly fits remaining space
+    beam2 = Beam(frame=Frame.worldXY(), length=4000, width=60, height=120)
+    assert stock.can_fit_beam(beam2)
+
+    # Add the beam
+    stock.add_beam(beam2)
+    assert stock.waste == 0
+
+    # Test beam that doesn't fit
+    beam3 = Beam(frame=Frame.worldXY(), length=1000, width=60, height=120)
+    assert not stock.can_fit_beam(beam3)
+
+
+def test_serialization():
+    """Test data serialization and copy functionality."""
+    # Create stock with beams
+    stock = Stock(6000, (60, 120))
+    beam1 = Beam(frame=Frame.worldXY(), length=2000, width=60, height=120)
+    beam2 = Beam(frame=Frame.worldXY(), length=1500, width=60, height=120)
+    stock.add_beam(beam1)
+    stock.add_beam(beam2)
+
+    # Set cutting_tolerance
+    stock.cutting_tolerance = 5.0
+
+    # Test serialization
+    data = stock.__data__
```

Comment:
better use `json_dumps` and `json_loads` otherwise we might be missing nested attributes that e.g. can't be (de)serialized